bundle/edu.gemini.pit/src/main/scala/edu/gemini/pit/ui/editor/ObservationEditor.scala

```diff
@@ -29,59 +31,115 @@ class ObservationEditor private (obs:Observation, canEdit:Boolean) extends StdMo
       case Some(_:NonSiderealTarget) => ICON_NONSIDEREAL
       case _                         => ICON_SIDEREAL
-    }))
+    }), gw=2)
     addSpacer()
-    addRow(new Label("Time"), new BorderPanel {
-      border = null
-      add(Time, BorderPanel.Position.Center)
-      add(Units, BorderPanel.Position.East)
-    })
-    preferredSize = (400, preferredSize.height) // force width
+    addLabeledRow(new RightLabel("Integration Time"), IntegrationTime, Units)
+    addLabeledRow(new RightLabel("Program Time"), ProgramTime, ProgramTimeUnits)
+    addLabeledRow(new RightLabel("Night Basecal Time"), PartTime, PartTimeUnits)
+    addLabeledRow(new RightLabel("Total Time"), TotalTime, TotalTimeUnits)
+    preferredSize = (500, preferredSize.height) // force width
   }
 
   // Editable
   Contents.Footer.OkButton.enabled = canEdit
 
   // Validation
-  override def editorValid = Time.valid
-  Time.reactions += {
+  override def editorValid = IntegrationTime.valid
+  IntegrationTime.reactions += {
     case ValueChanged(_) => validateEditor()
   }
 
-  type Named = {
-    def name:String
-  }
-
-  class OptionLabel(a:Option[Named], icon0:Icon) extends Label() {
-    horizontalAlignment = Alignment.Left
-    icon = icon0
-    text = a.map(_.name).getOrElse("none")
-    if (a.isEmpty)
-      foreground = Color.GRAY
-  }
+  // Time calculator
+  val calculator = obs.blueprint.flatMap(Overheads)
 
-  object Time extends NumberField(obs.time.map(_.value).orElse(Some(1.0)), allowEmpty = false) {
+  object IntegrationTime extends NumberField(obs.intTime.map(_.value).orElse(Some(1.0)), allowEmpty = false) {
     enabled = canEdit
     override def valid(d:Double) = d > 0
   }
 
   object Units extends ComboBox(TimeUnit.values.toList) with ValueRenderer[TimeUnit] {
     enabled = canEdit
-    selection.item = obs.time.getOrElse(TimeAmount.empty).units
+    selection.item = obs.intTime.getOrElse(TimeAmount.empty).units
+  }
+
+  class UnitsLabel extends Label {
+    def update(t: String): Unit =
+      text = t
+  }
+
+  class CalculatedTimeAmountField extends NumberField(None, allowEmpty = false) {
+    enabled = false
+
+    def update(t: TimeAmount): Unit =
+      value = (Units.selection.item match {
+        case TimeUnit.HR    => t.toHours
+        case TimeUnit.NIGHT => t.toNights
+      }).value
+  }
+
+  object ProgramTime extends CalculatedTimeAmountField
+  object ProgramTimeUnits extends UnitsLabel
+
+  object PartTime extends CalculatedTimeAmountField
+  object PartTimeUnits extends UnitsLabel
+
+  object TotalTime extends CalculatedTimeAmountField
+  object TotalTimeUnits extends UnitsLabel
+
+  def updateTimeLabels(): Unit = {
+    val intTime = TimeAmount(\/.fromTryCatchNonFatal(IntegrationTime.text.toDouble).getOrElse(0.0), Units.selection.item)
+    val obsTimes = calculator.map(_.calculate(intTime))
+    ProgramTime.update(obsTimes.map(_.progTime).getOrElse(TimeAmount.empty))
```

Comment:
I would be ok defining an additive `Monoid[TimeAmount]` even though it's not associative because it will never affect us at the scales we use. This would allow you to write this as `.update(obsTimes.foldMap(_.progTime))`